homeassistant/components/sensor/alpha_vantage.py

```diff
@@ -80,17 +81,20 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         _LOGGER.warning(msg)
         return
 
-    timeseries = TimeSeries(key=api_key)
-
     dev = []
+
+    timeseries = TimeSeries(key=api_key)
     for symbol in symbols:
         try:
             _LOGGER.debug("Configuring timeseries for symbols: %s",
                           symbol[CONF_SYMBOL])
             timeseries.get_intraday(symbol[CONF_SYMBOL])
```

Comment:
Is it necessary to do I/O already here?

Reply:
This was in the original source already. So I wouldn't know. I only move these lines below `dev = []` in line with the rest of the code further down.

Reply:
Ok, I don't know either, but having a sleep inside the `setup` is not acceptable (as it will make everything else in homeassistant to wait, too). Maybe it's fine to do an initial fetching here, and then just throttle the updates :+1:

Reply:
I fully understand that a sleep isn't the most elegant way to do this, but I can't think of another way taking into account how this API is working (accepting only an unknown but very lownumber of calls per second).

Reply:
Hi,
i wrote https://github.com/home-assistant/home-assistant/pull/15258 which also contains sleep statements, although in a more random fashion.
In order to ban the sleeps from the setup method, i think it is necessary to have only one query against the api. The current implementation runs a query for every configured stock/currencyexchange. When you simply want to check if the api key is valid, one query is enough.
Or not running queries in Setup() at all, because later in the update() methods one get's crash-log-entries anyway when the key is bad.

Reply:
When we don't make api calls at setup, a sensor will always be created regardless if a valid symbol is used or not. How would you reflect those errors on updates? Keep the state of the sensor at 'None' and send a 'critical' error?

Reply:
That would be the negative effect, that there can be a certain amount error messages, if you have many stocks to query. On the other side, if the api key is incorrect, then it should be corrected asap. If a symbol is incorrect, i also find it ok, if a system "screams", because it also needs to be fixed. The error messages an be switched off by simply maintaining a proper configuration.
And yes, returning None will make the symbols disappear and then the user will definively have a deeper look on what's wrong.